### ISSUE TITLE
[pytorch][bug] memory leaking for gloo backend when `all_reduce` CPU tensor

### ISSUE THREAD

**Issue by Stonesjtu** [pytorch/pytorch#7702]
The `dist.all_reduce` caused memory leaking when using gloo backend for CPU memory.
observation:
1. both `tcp` and `gloo` backends are tested, `tcp` is fine.
2. if `index` is of fixed size, it's fine.
3. other collective communication methods are not tested yet.
4. `reduce` tensors on GPU memory seems fine.
5. the memory usage increases linearly with the tensor size.

I think the leaking may happen in the buffer management of gloo backend.

reproducing snippets:
```python
# test.py
import time
import random

import torch
import torch.distributed as dist
def print_mem(extra_str=''):
    if dist.get_rank() == 0:
        import os
        import psutil
        process = psutil.Process(os.getpid())
        print(extra_str, process.memory_info().rss // 2**20, 'MB')

dist.init_process_group('gloo', world_size=2, init_method='file:///tmp/shared_file')
embedding = torch.randn(8000, 200)
for epo in range(1000):
    index = torch.randint(embedding.size(0), (30000 - epo,)).long()
    # ten = torch.randn(numel)
    ten = embedding[index]
    print_mem('before reducing')
    dist.all_reduce(ten, op=dist.reduce_op.SUM)
    print_mem('after reducing')
```

how to run:
```bash
pip install psutil
python test.py & python test.py
```

Envs
```
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: CentOS Linux 7 (Core)
GCC version: (GCC) 4.8.5
CMake version: version 3.6.3

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 8.0.61
GPU models and configuration:
GPU 0: GeForce GTX TITAN X
GPU 1: GeForce GTX TITAN X
GPU 2: GeForce GTX TITAN X
GPU 3: GeForce GTX TITAN X

Nvidia driver version: 375.26
cuDNN version: Probably one of the following:
/usr/local/cuda-7.5/lib64/libcudnn.so.5.0.5
/usr/local/cuda-7.5/lib64/libcudnn_static.a
/usr/local/cuda-8.0/lib64/libcudnn.so.5.1.5
/usr/local/cuda-8.0/lib64/libcudnn.so.6
/usr/local/cuda-8.0/lib64/libcudnn.so.6.0.21
/usr/local/cuda-8.0/lib64/libcudnn_static.a

Versions of relevant libraries:
[pip3] numpy (1.14.1)
[pip3] torch (0.4.0)
[conda] magma-cuda80              2.1.0                         5    soumith
[conda] torch-0.1.12              2                         <pip>
```

**Comment by zou3519** [comment]
cc @teng-li

**Comment by teng-li** [comment]
Old torch.distributed is deprecated and we will not support it as the highest priority,  can you try the latest version to see it is still the same problem?

**Comment by stgzr** [comment]
@teng-li I checked `all_gather`, `gather` and `all_reduce` using `gloo` backend on PyTorch v1.0. Maybe there is a memory leak in `gloo`? I also report a [bug](https://github.com/pytorch/pytorch/issues/16144)

**Comment by pietern** [comment]
Closing issue because it has gone stale. The issue reported by @stgzr has been fixed and released in 1.0.1.
